pkg/receive/handler.go

```diff
@@ -193,3 +212,80 @@ func ValidateLabels(logger log.Logger, next http.Handler, labels ...string) http
 		next.ServeHTTP(w, r)
 	}
 }
+
+func (h *Handler) TransformWriteRequest(logger log.Logger, next http.Handler) http.HandlerFunc {
+	logger = log.With(logger, "component", "receive", "middleware", "transformWriteRequest")
+	return func(w http.ResponseWriter, r *http.Request) {
+		body, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			level.Error(logger).Log("msg", "failed to read body", "err", err)
+			http.Error(w, "failed to read body", http.StatusInternalServerError)
+			return
+		}
+		defer r.Body.Close()
```

Comment:
Shouldn't this be moved before the `if`, to ensure we always close even on read error?

Reply:
Yeah, I copied this from the `LimitBodySize` middleware. Have replaced in all places!